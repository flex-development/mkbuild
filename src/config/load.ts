/**
 * @file Config - loadBuildConfig
 * @module mkbuild/config/loadBuildConfig
 */

import type { Config } from '#src/interfaces'
import { cosmiconfig } from 'cosmiconfig'
import es from './loader-es'

/**
 * Searches for a build configuration file.
 *
 * If found, a config object will be returned. The object will contain options
 * from the config file. If not found, an empty object will be returned.
 *
 * @see https://github.com/davidtheclark/cosmiconfig#explorersearch
 *
 * @async
 *
 * @param {string} [location] - Directory to search
 * @return {Config} Build configuration options
 */
const loadBuildConfig = async (location: string): Promise<Config> => {
  /**
   * Module name.
   *
   * @const {string} name
   */
  const name: string = 'build'

  // get search function
  const { search } = cosmiconfig(name, {
    loaders: { '.cts': es, '.js': es, '.mjs': es, '.mts': es, '.ts': es },
    searchPlaces: [
      '#.config.cjs',
      '#.config.cts',
      '#.config.js',
      '#.config.json',
      '#.config.mjs',
      '#.config.mts',
      '#.config.ts'
    ].map(place => place.replace('#', name)),
    stopDir: location
  })

  return ((await search(location))?.config as Config | null) ?? {}
}

export default loadBuildConfig
