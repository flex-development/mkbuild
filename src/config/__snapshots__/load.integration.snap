// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:config/loadBuildConfig > cosmiconfig > should load config from .cjs file 1`] = `
{
  "cwd": "__fixtures__/find-uniq",
  "ext": ".cjs",
  "format": "cjs",
  "pattern": "find-uniq.cts",
  "source": ".",
  "sourcemap": true,
  "sourcesContent": false,
}
`;

exports[`integration:config/loadBuildConfig > cosmiconfig > should load config from .cts file 1`] = `
{
  "bundle": true,
  "cwd": "__fixtures__/pkg/my-atoi",
  "ext": ".js",
  "format": "cjs",
  "name": "index",
  "platform": "neutral",
  "source": "find-uniq.cts",
  "sourcemap": true,
  "sourcesContent": false,
  "tsconfig": "tsconfig.json",
}
`;

exports[`integration:config/loadBuildConfig > cosmiconfig > should load config from .js file 1`] = `
{
  "bundle": true,
  "cwd": "__fixtures__/pkg/buddy",
  "entries": [
    {
      "format": "esm",
    },
    {
      "format": "cjs",
    },
  ],
  "name": "index",
  "platform": "neutral",
  "source": "buddy.js",
  "sourcemap": true,
  "sourcesContent": false,
}
`;

exports[`integration:config/loadBuildConfig > cosmiconfig > should load config from .json file 1`] = `
{
  "cwd": "__fixtures__/pkg/tribonacci",
  "entries": [
    {
      "format": "esm",
    },
    {
      "format": "cjs",
    },
  ],
  "sourcemap": true,
  "sourcesContent": false,
  "tsconfig": "tsconfig.json",
}
`;

exports[`integration:config/loadBuildConfig > cosmiconfig > should load config from .mjs file 1`] = `
{
  "cwd": "__fixtures__/pkg/sum-of-intervals",
  "entries": [
    {
      "format": "esm",
      "pattern": "sum-of-intervals.mjs",
    },
    {
      "format": "cjs",
      "pattern": "sum-of-intervals.cjs",
    },
  ],
  "source": ".",
  "sourcemap": true,
  "sourcesContent": false,
}
`;

exports[`integration:config/loadBuildConfig > cosmiconfig > should load config from .mts file 1`] = `
{
  "conditions": Set {
    "node",
    "import",
  },
  "cwd": "__fixtures__/pkg/reverse",
  "sourcemap": true,
  "sourcesContent": false,
  "tsconfig": "tsconfig.json",
}
`;

exports[`integration:config/loadBuildConfig > cosmiconfig > should load config from .ts file 1`] = `
{
  "cwd": "__fixtures__/pkg/dbl-linear",
  "format": "esm",
  "outdir": ".",
  "pattern": "dbl-linear.ts",
  "source": ".",
  "sourcemap": true,
  "sourcesContent": false,
  "tsconfig": "tsconfig.json",
}
`;
